stl/src/vector_algorithms.cpp

```diff
@@ -1952,29 +1983,39 @@ namespace {
         size_t _Result = 0;
 
 #ifndef _M_ARM64EC
-        size_t _Size_bytes = _Byte_length(_First, _Last);
+        const size_t _Size_bytes = _Byte_length(_First, _Last);
 
-        const size_t _Avx_size = _Size_bytes & ~size_t{0x1F};
-        if (_Avx_size != 0 && _Use_avx2()) {
+        if (const size_t _Avx_size = _Size_bytes & ~size_t{0x1F}; _Avx_size != 0 && _Use_avx2()) {
             const __m256i _Comparand = _Traits::_Set_avx(_Val);
             const void* _Stop_at     = _First;
             _Advance_bytes(_Stop_at, _Avx_size);
+
             do {
                 const __m256i _Data = _mm256_loadu_si256(static_cast<const __m256i*>(_First));
                 const int _Bingo    = _mm256_movemask_epi8(_Traits::_Cmp_avx(_Data, _Comparand));
                 _Result += __popcnt(_Bingo); // Assume available with SSE4.2
                 _Advance_bytes(_First, 32);
             } while (_First != _Stop_at);
-            _Size_bytes &= 0x1F;
+
+            if (const size_t _Avx_tail_size = _Size_bytes & 0x1C; _Avx_tail_size != 0) {
+                const __m256i _Tail_mask = _Avx2_tail_mask_32(_Avx_tail_size >> 2);
+                const __m256i _Data      = _mm256_maskload_epi32(static_cast<const int*>(_First), _Tail_mask);
+                const int _Bingo =
+                    _mm256_movemask_epi8(_mm256_and_si256(_Traits::_Cmp_avx(_Data, _Comparand), _Tail_mask));
+                _Result += __popcnt(_Bingo); // Assume available with SSE4.2
+                _Advance_bytes(_First, _Avx_tail_size);
+            }
 
             _mm256_zeroupper(); // TRANSITION, DevCom-10331414
-        }
 
-        const size_t _Sse_size = _Size_bytes & ~size_t{0xF};
-        if (_Sse_size != 0 && _Use_sse42()) {
+            if constexpr (sizeof(_Ty) >= 4) {
+                return _Result >>= _Traits::_Shift;

```

Comment:
This caught my attention because using a compound assignment to return a plain value is unnecessary.

After considering this, I think what's actually strange is where we're mentioning `_Traits::_Shift` now. Previously (before this PR), it was mentioned exactly once which was fine, even though it (somewhat unusually) occurred outside of the `_M_ARM64EC` vectorized region so that it could be used for all fallthrough codepaths. Now, `_Traits::_Shift` is being mentioned twice, once in this early return and again later.

There isn't a *whole* lot of potential for confusion here, but I think it would be clearer if we instead moved `_Result >>= _Traits::_Shift;` to the end of the (now separate) AVX2 and SSE4.2 codepaths, which will make it clearer that (1) this is for vectorization only and has no effect when we don't have the instructions and (2) this is needed when we're getting ready to return (either early, or going through the fallback at the bottom).

Reply:
Yes, this is consistent with the same situation in newly implemented `mismatch`:
https://github.com/microsoft/STL/blob/56ad8253a8efe3bbb5bbde42c9fd755764959f44/stl/src/vector_algorithms.cpp#L2190-L2195
`mismatch` has `else` from the beginning, so was initially implemented with count adjustment in the branches.